### ISSUE TITLE
integrate_load_illegal_guest_permissions

### ISSUE THREAD

**Issue by cody3647** [elkarte/Elkarte#3317]
All the hook seems to do is load the permissions in $context['non_guest_permission'] but not actually into the Permissions Class illegal_guest_permissions property.  Not sure what it does without doing that, everywhere else seems to use the property not $context.

https://github.com/elkarte/Elkarte/blob/7b3278f1ca8646292f6056c506bf341daec1c370/sources/subs/Permissions.class.php#L126

**Comment by Spuds** [comment]
Hummm ... looks like it needs a ```$this->illegal_guest_permissions = $context['non_guest_permissions'];``` after the integration hook.   

A similar issue is in ```loadIllegal``` where it does a ```$this->illegal_permissions = $illegal_permissions;``` and I think that should be ```$this->illegal_permissions = $context['illegal_permissions'];``` instead.

@emanuele45 what do you think?

**Comment by cody3647** [comment]
Well currently $context['non_guest_permissions'] isn't used anywhere but internally in the Permissions class.  And in the ManageDraftsModule controller where it uses the hook.

As far as I can tell, getIllegalPermissions() and getIllegalGuestPermissions() always load into a local variable.

**Comment by emanuele45** [comment]
I guess that technically $context['non_guest_permissions'] was left for backward compatibility (and in 1.1 I would agree with it).
Indeed it looks broken.
Either what you suggested @Spuds or 
```
$this->illegal_guest_permissions = &$context['non_guest_permissions'];
```
just to avoid one more line, even though it's a bit meh. :man_shrugging: 

I would also consider to pass the variable to the hook for consistency and with the goal of removing the global in 2.0?

```
$context['non_guest_permissions'] = $this->illegal_guest_permissions;
		call_integration_hook('integrate_load_illegal_guest_permissions', [&$this->illegal_guest_permissions]);
```

What do you think?

**Comment by Spuds** [comment]
For completeness a simple unit test as well https://github.com/elkarte/Elkarte/commit/9b5f57f6cb18318c210d000a72dde2aaa5bf4266

**Comment by emanuele45** [comment]
I passed protected properties all the time. :innocent: 
Well, to me it at least for the time being it makes sense, otherwise we'd have to pass the whole object and have methods to alter these properties...
But that would mostly break any backward compatibility (and I think even SP uses them).
The problem I see with assign $this->illegal_guest_permissions to $context['non_guest_permissions'] after the hook is that *right now* most of the addons using this hook will alter $context['non_guest_permissions'], and this change will obliterate it. And at the moment will likely cause also some undefined index warnings, because if the addons hook is like this:
```
function myaddon_hook()
{
	global $context;
	$context['illegal_permissions'][] = 'my_permission';
}
```
$context['illegal_permissions'] is likely not defined causing an error.

If we don't want to disrupt addons behaviour I think the two options are:
```
$context['illegal_permissions'] = $this->illegal_permissions;
call_integration_hook('integrate_load_illegal_permissions', array(&$this->illegal_permissions));
$context['illegal_permissions'] = array_merge($context['illegal_permissions'], $this->illegal_permissions); // or the other way around depending on... reasons
```
or:
```
$context['illegal_permissions'] = &$this->illegal_permissions;
call_integration_hook('integrate_load_illegal_permissions', array(&$this->illegal_permissions));
```

**Comment by Spuds** [comment]
I'm somehow surprised that php allows that ... oh wait, no I'm not 😛 

I had not considered the mixed environment of addons using the global and/or the passed param, good catch.   Here is another attempt https://github.com/Spuds/Elkarte/commit/b304c5cc49d425a29da809013a67c7746c0b47d5

**Comment by emanuele45** [comment]
You mean I slipped this through your radar?
https://github.com/elkarte/Elkarte/blob/7b3278f1ca8646292f6056c506bf341daec1c370/sources/subs/EventManager.class.php#L100
https://github.com/elkarte/Elkarte/blob/7b3278f1ca8646292f6056c506bf341daec1c370/sources/Action.controller.php#L277

:grin: 
With events and triggers an addon can access almost anything belonging to the controller. :angel: 
It's not really nice, I agree, but when I did it my reasoning was that we are used to (and some times we *have to*) access internals, and if we don't use some kind of "hook" we use modifications (that are way worse), so I considered an acceptable evil.
